File: LeetCode/Easy/Cousins In A Binary Tree/Go/cousins.go
package main

import "fmt"

type TreeNode struct {
	Val int
	Left,
	Right *TreeNode
}

func New(val int) *TreeNode { return &TreeNode{Val: val, Left: nil, Right: nil} }

type Pair struct {
	node   *TreeNode
	depth  int
	parent int
}

func cousins(root *TreeNode, x, y int) bool {
	if root == nil {
		return false
	}
	q := []Pair{{node: root, depth: 0, parent: -1}}
	m, n := Pair{}, Pair{}
	for len(q) > 0 {
		curr := q[0].node
		depth := q[0].depth
		parent := q[0].parent
		q = q[1:]
		if curr.Val == x {
			m.parent = parent
			m.depth = depth
			m.node = curr
		}

		if curr.Val == y {
			n.parent = parent
			n.depth = depth
			n.node = curr
		}

		if curr.Left != nil {
			q = append(q, Pair{node: curr.Left, depth: depth + 1, parent: curr.Val})
		}

		if curr.Right != nil {
			q = append(q, Pair{node: curr.Right, depth: depth + 1, parent: curr.Val})
		}
	}

	if m.depth == n.depth {
		return m.parent != n.parent
	} else {
		return false
	}
}

func main() {
	root := New(1)
	root.Left = New(2)
	root.Right = New(3)
	root.Left.Left = New(4)
	fmt.Println(cousins(root, 3, 4))
}

File: LeetCode/Easy/Cousins In A Binary Tree/c++/cousins.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isCousins(TreeNode* root, int x, int y) {
        queue<pair<int, TreeNode*>> q;
        q.push({-1, root});

        while (!q.empty()) {
            int size = q.size();
            int parent_x = 0;
            int parent_y = 0;
            bool found_x = false;
            bool found_y = false;
            for (int i = 0; i < size; i++) {
                int parent_val = q.front().first;
                TreeNode* curr = q.front().second;
                q.pop();

                if (x == curr->val || y == curr->val && parent_val == -1) return false;
                
                if (curr->left != nullptr) {
                    q.push({curr->val, curr->left});
                    if (curr->left->val == x) {
                        parent_x = curr->val;
                        found_x = true;
                    }
                    if (curr->left->val == y) {
                        parent_y = curr->val;
                        found_y = true;
                    }
                }
                
                if (curr->right != nullptr) {
                    q.push({curr->val, curr->right});
                    if (curr->right->val == x) {
                        parent_x = curr->val;
                        found_x = true;
                    }
                    if (curr->right->val == y) {
                        parent_y = curr->val;
                        found_y = true;
                    }
                }
                
                if (found_x && found_y && parent_x == parent_y) return false;
                if (found_x && found_y) return true;
            }
        }
        return false;
    }
};
